### PR TITLE
Minimum camera zoom distance for spring and overhead cameras

### PR DESCRIPTION
**Pull request by loveridge**
fixes https://github.com/beyond-all-reason/spring/issues/1548

### CODE REVIEW

**Comment by badosu** [comment]
At a brief glance this looks good to me, unfortunately I won't be able to test and review this more deeply. I'll await someone else for approval. Thanks for the contribution!

A note: The spring cam minimum height is distance to ground when tracing a ray in the frustum direction instead of ground normal tangent to frustum location right (what would normally be interpreted as height)? I might be incorrect so let me know, regardless the change makes sense to me since the dist is the canonical reference for this cam mode, but it's good gamedevs are aware of if the quirk exists (add to the mincameraheight description).

File: rts/Game/Camera/SpringController.cpp [hunk 1]
```diff
@@ -258,8 +261,11 @@ float CSpringController::ZoomIn(const float3& curCamPos, const float3& newDir, c
 	const float3 wantedPos = curCamPos + cursorVec * (1.0f - scaledMode);
 
 	// figure out how far we will end up from the ground at new wanted point
-	curDist = DistanceToGround(wantedPos, dir, pos.y);
-	pos = wantedPos + dir * curDist;
+	float wantedDist = DistanceToGround(wantedPos, dir, pos.y);
+	if (wantedDist > minDist) {
```
Comment (badosu) [inline rts/Game/Camera/SpringController.cpp:265]: Maybe this could get into a situation where the -delta makes this condition not pass (e.g. very high scroll delta on single tick), but you're not close to minheight. Perhaps clamp wantedDist instead?

File: rts/Game/Camera/OverheadController.cpp [hunk 2]
```diff
@@ -158,7 +161,7 @@ void COverheadController::MouseWheelMove(float move, const float3& newDir)
 			if ((wantedPos.y + (dir.y * newHeight)) < 0.0f)
 				newHeight = -wantedPos.y / yDirClamp;
 
-			if (newHeight < maxHeight) {
+			if (newHeight < maxHeight && newHeight > minHeight) {
```
Comment (badosu) [inline rts/Game/Camera/OverheadController.cpp:164]: See like 146, there's some kind of clamping there. We might want to use minHeight instead of harcoded 60.0

Comment (badosu) [reply rts/Game/Camera/OverheadController.cpp:164]: On second thought this is relevant but perhaps out of scope for this PR, I created a separate issue: https://github.com/beyond-all-reason/spring/issues/1578

**Comment by Beherith** [comment]
Having slammed against this a few times already, be sure to check the really wierd edge cases, where:
1. Camera is outside of map bounds
2. Camera is lookup upwards from horizontal
3. Camera is looking straight ahead
4. There is some very steep and varying terrain nearby. 

I'll gladly test .exe's, but dont have a toolchain at the moment.

**Comment by sprunk** [comment]
5. camera is underwater (surface vs seafloor)

File: rts/Game/Camera/SpringController.cpp [hunk 3]
```diff
@@ -73,6 +75,7 @@ void CSpringController::ConfigUpdate()
 	RECOIL_DETAILED_TRACY_ZONE;
 	scrollSpeed = configHandler->GetFloat("CamSpringScrollSpeed") * 0.1f;
 	fov = configHandler->GetFloat("CamSpringFOV");
+	minDist = configHandler->GetFloat("CamSpringMinHeight");
```
Comment (dzosz) [inline rts/Game/Camera/SpringController.cpp:78]: I think this approach will not work consistently and will get worse when camera angle is flatter. `pos` in spring controller is a focus point on the ground, `dist` is a distance from that point using current rotation (`dir`), actual camera position is calculated in method `GetPos()`. There's some code that limits/manipulates camera height in method `SmoothCamHeight()`  by changing point of focus (`pos`). Maybe it can be reused?

Comment (loveridge) [reply rts/Game/Camera/SpringController.cpp:78]: What overhead camera calls height is actually zoom distance, and I carried over the name here to spring camera. I will rename both to zoom distance.

Comment (badosu) [reply rts/Game/Camera/SpringController.cpp:78]: I feel their concern is mostly about the experience of the limit than the name itself. That's the point I raised in my note, is the dist checked against minDist the ray traced from frustum position into dir or the actual height to ground?

Comment (badosu) [reply rts/Game/Camera/SpringController.cpp:78]: e.g. your actual height might be below mindist but looking at the horizon the check might not work

Comment (loveridge) [reply rts/Game/Camera/SpringController.cpp]: All of these "heights" are rays out of the camera frustum, therefore I will rename the config values to make it clear.

**Comment by loveridge** [comment]
I've fixed the issue with large zoom scroll speeds, and renamed the configs to minimum zoom distance rather than "height".

**Comment by badosu** [comment]
As I mentioned in my original comment it would be very good to clarify what the height for spring camera means, if it is the distance to traceray to ground then it needs more details on the description so we get it clear now instead of problems in the future. If it's not it's all good

**Comment by loveridge** [comment]
> As I mentioned in my original comment it would be very good to clarify what the height for spring camera means, if it is the distance to traceray to ground then it needs more details on the description so we get it clear now instead of problems in the future. If it's not it's all good

It is a traceray to ground along the camera's direction vector for both cameras. So more of a "minimum zoom distance", rather than a "camera height".
image

**Comment by badosu** [comment]
Yeah, so the action item is to add that to the description, since the setting has height in the name it might confuse game developers

**Comment by loveridge** [comment]
> Yeah, so the action item is to add that to the description, since both spring and overhead cams have zoom in the name, with different behaviors, it might confuse game developers

I changed both config names to zoom distance, as they both represent the minimum value you can zoom in, and not the minimum height from the camera to the ground on the Y axis.
```c++
CONFIG(float, OverheadMinZoomDistance).defaultValue(60.0f).description("Minimum camera zoom distance");

CONFIG(float, CamSpringMinZoomDistance).defaultValue(20.0f).description("Minimum camera zoom distance");
```

**Comment by badosu** [comment]
That's alright, I can do it myself I guess. Thank you!
